File: src/Billing/CartService.php
<?php
declare(strict_types=1);
namespace App\Billing;
use App\Infrastructure\Database;
final class CartService
{
    public function __construct(private Database $db) {}
    /** Save cart data. When $intent is set, a topup may auto-purchase the cart. */
    public function save(string $userId, array $data, bool $intent = false): void
    {
        $json = json_encode($data, JSON_THROW_ON_ERROR);
        $this->db->execute(
            'INSERT INTO carts(user_id,data,intent,updated_at) VALUES(?,?,?,?) ON CONFLICT(user_id) DO UPDATE SET data=excluded.data,intent=excluded.intent,updated_at=excluded.updated_at',
            [$userId, $json, (int)$intent, time()]
        );
    }
    public function get(string $userId): ?array
    {
        $row = $this->db->one('SELECT data,intent FROM carts WHERE user_id = ?', [$userId]);
        if (!$row) return null;
        $data = json_decode($row['data'], true, 512, JSON_THROW_ON_ERROR);
        $data['_intent'] = (int)$row['intent'] === 1;
        return $data;
    }
    public function delete(string $userId): void
    {
        $this->db->execute('DELETE FROM carts WHERE user_id = ?', [$userId]);
    }
    public function hasIntent(string $userId): bool
    {
        return (int)($this->db->one('SELECT intent FROM carts WHERE user_id = ?', [$userId])['intent'] ?? 0) === 1;
    }
    public function clearIntent(string $userId): void
    {
        $this->db->execute('UPDATE carts SET intent = 0 WHERE user_id = ?', [$userId]);
    }
}

File: ruby_app/zeleboba/billing/cart_service.rb
# frozen_string_literal: true

require "json"

require_relative "error"

module Zeleboba
  module Billing
    class CartService
      def initialize(db)
        @db = db
      end

      def save(user_id, data, intent: false)
        raise Error, "Некорректная корзина." unless data.is_a?(Hash)
        payload = data.transform_keys(&:to_s).slice("kind", "plan_id", "subscription_id", "traffic_gb", "devices", "price_kopeks", "recipient_type", "recipient_value", "message", "idempotency_key")
        @db.execute(
          "INSERT INTO carts(user_id,data,intent,updated_at) VALUES(?,?,?,?) ON CONFLICT(user_id) DO UPDATE SET data=excluded.data,intent=excluded.intent,updated_at=excluded.updated_at",
          [user_id, JSON.generate(payload), intent ? 1 : 0, Time.now.to_i]
        )
      end

      def fetch(user_id)
        row = @db.one("SELECT * FROM carts WHERE user_id=?", [user_id])
        return nil unless row

        JSON.parse(row["data"]).merge("_intent" => row["intent"].to_i == 1)
      rescue JSON::ParserError
        raise Error, "Корзина повреждена. Создайте покупку заново."
      end

      def clear_intent(user_id)
        @db.execute("UPDATE carts SET intent=0,updated_at=? WHERE user_id=?", [Time.now.to_i, user_id])
      end

      def delete(user_id)
        @db.execute("DELETE FROM carts WHERE user_id=?", [user_id])
      end
    end
  end
end
